packages/shared/src/components/modals/SlackIntegrationModal.tsx

```diff
@@ -0,0 +1,351 @@
+  source,
+  ...props
+}: SlackIntegrationModalProps): ReactElement => {
+  const isMobile = useViewSize(ViewSize.MobileL);
+  const { removeSourceIntegration } = useIntegration();
+  const queryClient = useQueryClient();
+  const { user } = useAuthContext();
+  const { displayToast } = useToastNotification();
+
+  const [state, setState] = useState<{
+    userIntegration?: UserIntegration;
+    channelId?: string;
+  }>({});
+
+  const { data: sourceIntegration } = useSourceIntegration({
+    userIntegrationType: UserIntegrationType.Slack,
+    sourceId: source.id,
+  });
+
+  const { data: slackIntegrations, isLoading: isLoadingIntegrations } =
+    useIntegrations({
+      queryOptions: {
+        select: useCallback((data) => {
+          const filteredData = data.filter(
+            (integration) => integration.type === UserIntegrationType.Slack,
+          );
+
+          if (filteredData.length > 0) {
+            filteredData.push({
+              id: 'new',
+              type: UserIntegrationType.Slack,
+              createdAt: new Date(),
+              updatedAt: new Date(),
+              name: 'Connect another workspace',
+              userId: 'new',
+            });
+          }
+
+          return filteredData;
+        }, []),
+      },
+    });
+
+  const selectedIntegration =
+    state.userIntegration ||
+    sourceIntegration?.userIntegration ||
+    slackIntegrations?.[0];
+  const selectedIntegrationIndex = slackIntegrations?.findIndex(
+    (item) => item.id === selectedIntegration?.id,
+  );
+
+  const { data: channels } = useSlackChannels({
+    integrationId: selectedIntegration?.id,
+  });
+
+  const selectedChannel = state.channelId || sourceIntegration?.channelIds[0];
+  const selectedChannelIndex = channels?.findIndex(
+    (item) => item.id === selectedChannel,
+  );
+
+  const slack = useSlack();
+
+  const onConnectNew = () => {
+    slack.connect({
+      redirectPath: `/${
+        source.type === SourceType.Squad ? 'squads' : 'sources'
+      }/${source.handle}?lzym=${LazyModal.SlackIntegration}`,
+    });
+  };
+
+  const { mutateAsync: onSave, isLoading: isSaving } = useMutation(
+    async () => {
+      await slack.connectSource({
+        channelId: channels[selectedChannelIndex].id,
+        integrationId: selectedIntegration.id,
+        sourceId: source.id,
+      });
+    },
+    {
+      onSuccess: () => {
+        queryClient.invalidateQueries(
+          generateQueryKey(RequestKey.UserSourceIntegrations, user),
+        );
+
+        displayToast(labels.integrations.success.integrationSaved);
+      },
+    },
+  );
+
+  const noIntegrations = !slackIntegrations?.length && !isLoadingIntegrations;
+
+  const { slackIntegrationHeader } = useThemedAsset();
+
+  const modalProps: ModalProps = {
+    kind: Modal.Kind.FlexibleCenter,
+    size: Modal.Size.XSmall,
+    isDrawerOnMobile: true,
+    ...props,
+  };
+
+  if (isLoadingIntegrations) {
+    return (
+      <Modal {...modalProps}>
```

Comment:
Won't this flash? by showing/hiding a whole modal?

Reply:
it does not from what I see, but regarding your comment https://github.com/dailydotdev/apps/pull/3315#discussion_r1709628746 I will refactor the insides

Reply:
Refactored here 46ae06df59c3cda83080ad6b1bdfc273e8a6d3e4